qt: Update SetHexDeprecated to FromHex

marcofleon opened the pull request:
This is part of https://github.com/bitcoin/bitcoin/pull/32189. I'm separating this out because it's not immediately obvious that it's just a refactor. `SetHexDeprecated()` doesn't do any correctness checks on the input, while `FromHex()` does, so it's theoretically possible that there's a behavior change.

Replaces `uint256::SetHexDeprecated()` calls with `Txid::FromHex()` in four locations:
- `TransactionTableModel::updateTransaction`
- `TransactionView::contextualMenu`
- `TransactionView::abandonTx`
- `TransactionView::bumpFee`

The input strings in these cases aren't user input, so they should only be valid hex strings from `GetHex()` (through `TransactionRecord::getTxHash()`). These conversions should be safe without additional checks.

marcofleon committed:
qt: Update SetHexDeprecated to FromHex

Replace `uint256::SetHexDeprecated()` calls with `Txid::FromHex()`
in four locations:
- TransactionTableModel::updateTransaction
- TransactionView::contextualMenu
- TransactionView::abandonTx
- TransactionView::bumpFee

The input strings are generally expected to be valid hex strings
from `GetHex()`. However, due to the potentially unpredictable return
value of `.data(TransactionTableModel::TxHashRole)`, check the
`Txid::FromHex` result in `contextualMenu` and return early if the
transaction hash is invalid. The other two functions, `abandonTx`
and `bumpFee` will only be called if the context menu is enabled.

marcofleon commented:
@hebasto @furszy If you could look this over and make sure I didn't miss anything here, I would appreciate it.

instagibbs commented:
suggestion to maybe nuke that last usage of `SetHexDeprecated` and then just nuke it entirely?

laanwj commented on src/qt/transactiontablemodel.cpp:
```diff
@@ -276,8 +276,7 @@ void TransactionTableModel::updateAmountColumnTitle()
 {
-    uint256 updated;
-    updated.SetHexDeprecated(hash.toStdString());
+    Txid updated = Txid::FromHex(hash.toStdString()).value();
```
This seems reasonable: `updateTransaction` is called from `TransactionNotification::invoke`, which creates the string using `uint256::GetHex`.

hebasto replied:
This use of `SetHexDeprecated` could be eliminated by changing the type of the first parameter from `const QString&` to `const uint256&`. See [this commit](https://github.com/hebasto/bitcoin/commit/01b84ba9a78597d9f7e1ab0c76fda0c56b32980f) in [this branch](https://github.com/hebasto/bitcoin/commits/250409-pr32237-AMENDED/).

laanwj commented on src/qt/transactionview.cpp:
```diff
@@ -395,8 +395,8 @@ void TransactionView::contextualMenu(const QPoint &point)
-    uint256 hash;
-    hash.SetHexDeprecated(selection.at(0).data(TransactionTableModel::TxHashRole).toString().toStdString());
+    QString hashQStr = selection.at(0).data(TransactionTableModel::TxHashRole).toString();
+    Txid hash = Txid::FromHex(hashQStr.toStdString()).value();
```
The string here comes from `TransactionTableModel::data(TxHashRole)`, which is filled in from `TransactionRecord::getTxHash` which calls `uint256::GetHex`.

So far, so good, but because of potential unpredictability in internal Qt logic (eg what if the table has an incomplete row for some reason) i'm not 100% reassured that this can never return an empty string, so would prefer to err on the side of "don't display the context menu" to making an hard assumption.

marcofleon replied:
Thanks for reviewing. Let me know if you had something else in mind for this check. I think the early return (not showing the menu?) is a slight behavior change, as opposed to just disabling the actions. If that works then nice, but if not then I can change it to be exactly the old behavior.

laanwj commented on src/qt/transactionview.cpp:
```diff
@@ -414,9 +414,8 @@ void TransactionView::abandonTx()
-    uint256 hash;
     QString hashQStr = selection.at(0).data(TransactionTableModel::TxHashRole).toString();
-    hash.SetHexDeprecated(hashQStr.toStdString());
+    Txid hash = Txid::FromHex(hashQStr.toStdString()).value();
```
Seems fine, same reasoning about `.data()` as before. Plus if we get here, the context menu was enabled.

laanwj commented on src/qt/transactionview.cpp:
```diff
@@ -429,9 +428,8 @@ void TransactionView::bumpFee([[maybe_unused]] bool checked)
-    uint256 hash;
     QString hashQStr = selection.at(0).data(TransactionTableModel::TxHashRole).toString();
-    hash.SetHexDeprecated(hashQStr.toStdString());
+    Txid hash = Txid::FromHex(hashQStr.toStdString()).value();
```
Same case as `abandonTx()`, this is called from the context menu only.

laanwj reviewed (changes requested):
Changes LGTM, but would prefer handling `data(TxHashRole)` returning a null QVariant more robustly in `TransactionView::contextualMenu`.

laanwj commented:
> suggestion to maybe nuke that last usage of SetHexDeprecated and then just nuke it entirely?

nuke it

hebasto commented:
Concept ACK.

marcofleon committed:
refactor: Remove SetHexDeprecated

After replacing all instances of `SetHexDeprecated` in the GUI,
remove it entirely and reimplement the behavior in `FromHex`.

marcofleon commented:
Pushed changes addressing @instagibbs and @laanwj 's suggestions. Let me know if I missed something.

laanwj commented on src/uint256.h:
```diff
@@ -158,7 +144,16 @@ std::optional<uintN_t> FromHex(std::string_view str)
+    size_t digits = str.size();
+    while (digits > 0 && p1 < pend) {
+        *p1 = ::HexDigit(str[--digits]);
+        if (digits > 0) {
```
Because the `uintN_t::size() * 2 == str.size()` check already makes sure there can be no underrun here, one could shorten this to
```c++
*p1 = (unsigned char)::HexDigit(str[digits - 1]) | ((unsigned char)::HexDigit(str[digits - 2]) << 4);
digits -= 2;
```
That said, for a refactor it seems better to move the code as obviously as possible as done here.

laanwj reviewed (approved):
Code review ACK 868816d962ac55346381025e10bedc12dfcce113

w0xlt reviewed (commented):
Code review ACK https://github.com/bitcoin/bitcoin/pull/32237/commits/868816d962ac55346381025e10bedc12dfcce113

BrandonOdiwuor reviewed (commented):
Code Review ACK 868816d962ac55346381025e10bedc12dfcce113

sedited reviewed (approved):
ACK 868816d962ac55346381025e10bedc12dfcce113

hebasto reviewed (approved):
ACK 868816d962ac55346381025e10bedc12dfcce113, I have reviewed the code and it looks OK.
